src/app/tests/TestAttributePathExpandIterator.cpp

```diff
@@ -217,18 +328,45 @@ TEST(TestAttributePathExpandIterator, TestWildcardAttribute)
         { kMockEndpoint2, MockClusterId(3), Clusters::Globals::Attributes::AttributeList::Id },
     };
 
-    size_t index = 0;
-
-    for (app::AttributePathExpandIterator iter(CodegenDataModelProviderInstance(nullptr /* delegate */), &clusInfo); iter.Get(path);
-         iter.Next())
     {
-        ChipLogDetail(AppServer, "Visited Attribute: 0x%04X / " ChipLogFormatMEI " / " ChipLogFormatMEI, path.mEndpointId,
-                      ChipLogValueMEI(path.mClusterId), ChipLogValueMEI(path.mAttributeId));
-        EXPECT_LT(index, ArraySize(paths));
-        EXPECT_EQ(paths[index], path);
-        index++;
+        size_t index = 0;
+
+        auto position = AttributePathExpandIterator::Position::StartIterating(&clusInfo);
+        app::AttributePathExpandIterator iter(CodegenDataModelProviderInstance(nullptr /* delegate */), position);
+
+        while (iter.Next(path))
+        {
+            ChipLogDetail(AppServer, "Visited Attribute: 0x%04X / " ChipLogFormatMEI " / " ChipLogFormatMEI, path.mEndpointId,
+                          ChipLogValueMEI(path.mClusterId), ChipLogValueMEI(path.mAttributeId));
+            EXPECT_LT(index, ArraySize(paths));
+            EXPECT_EQ(paths[index], path);
+            index++;
+        }
+        EXPECT_EQ(index, ArraySize(paths));
+    }
+    // identical test, but re-using position and re-creating iterators
```

Comment:
I think the point is proven by just using this new structure added, rather than doing 2 tests, as the one where the iterator is re-created is the one which would fail.

Reply:
It seems indeed likely that if the split-out 2nd test passes,the 1st should as well. However I was not sure that it is a given in case the constructor has to do "fancy" initialization in its constructor (we are considering moving to getting spans/memory tables of metadata and that would happen in the constructor, making the construct vs on-going operation potentially different).

So even though 2nd pass makes 1st pass very likely, I still believe that this separation does give us better coverage/confidence in the code. Since it is reasonably fast to run, it seems more coverage even if reasonably overlapping is good.

Reply:
Updated most tests to not duplicate work and kept the "position resume" version. Kept the dual version in one of the tests only just in case.